### ISSUE TITLE
dtls1_process_record  logic seems conflict with RFC6347

### ISSUE THREAD

**Issue by KanPlus** [openssl/openssl#19062]
As Per RFC6347, invalid records  should be silently discarded, and generate fatal level alerts in case of generate an alert.  But dtls1_process_record have these two behavior.  This seems conflict with the RFC6347.

dtls1_process_record implement:
generate fatal level alerts:
````
        if (!ossl_assert(mac_size <= EVP_MAX_MD_SIZE)) {
            SSLfatal(s, SSL_AD_INTERNAL_ERROR, SSL_F_DTLS1_PROCESS_RECORD,
                     ERR_R_INTERNAL_ERROR);
            return 0;
        }
````
 generate an alert:
````
    if (enc_err < 0) {
        /* decryption failed, silently discard message */
        rr->length = 0;
        RECORD_LAYER_reset_packet_length(&s->rlayer);
        return 0;
    }
````

> [4.1.2.7](https://www.rfc-editor.org/rfc/rfc6347#section-4.1.2.7).  Handling Invalid Records
>    Unlike TLS, DTLS is resilient in the face of invalid records (e.g., invalid formatting, length, MAC, etc.).  In general, invalid records
>    SHOULD be silently discarded, thus preserving the association; however, an error MAY be logged for diagnostic purposes.   Implementations which choose to generate an alert instead, MUST  generate fatal level alerts to avoid attacks where the attacker    repeatedly probes the implementation to see how it responds to  various types of error.  Note that if DTLS is run over UDP, then any   implementation which does this will be extremely susceptible to denial-of-service (DoS) attacks because UDP forgery is so easy.  Thus, this practice is NOT RECOMMENDED for such transports. 
>      If DTLS is being carried over a transport that is resistant to  forgery (e.g., SCTP with SCTP-AUTH), then it is safer to send alerts because an attacker will have difficulty forging a datagram that will  not be rejected by the transport layer.

**Comment by mattcaswell** [comment]
````
        if (!ossl_assert(mac_size <= EVP_MAX_MD_SIZE)) {
            SSLfatal(s, SSL_AD_INTERNAL_ERROR, SSL_F_DTLS1_PROCESS_RECORD,
                     ERR_R_INTERNAL_ERROR);
            return 0;
        }
````
In this case this is a "should not happen" internal error. It is not the result of receiving an invalid record so the text about how to handle invalid records in 4.1.2.7 does not apply. Sending a fatal alert in this case is reasonable.

````
    if (enc_err < 0) {
        /* decryption failed, silently discard message */
        rr->length = 0;
        RECORD_LAYER_reset_packet_length(&s->rlayer);
        return 0;
    }
````
I'm not sure what your point is with this code. It seems to me to be doing exactly what the RFC says, i.e. we have encountered an invalid record and therefore we are silently discarding it. No alert should be generated in this case (otherwise it wouldn't be silent).

**Comment by KanPlus** [comment]
> ```
>         if (!ossl_assert(mac_size <= EVP_MAX_MD_SIZE)) {
>             SSLfatal(s, SSL_AD_INTERNAL_ERROR, SSL_F_DTLS1_PROCESS_RECORD,
>                      ERR_R_INTERNAL_ERROR);
>             return 0;
>         }
> ```
> 
> In this case this is a "should not happen" internal error. It is not the result of receiving an invalid record so the text about how to handle invalid records in 4.1.2.7 does not apply. Sending a fatal alert in this case is reasonable.
> 
> ```
>     if (enc_err < 0) {
>         /* decryption failed, silently discard message */
>         rr->length = 0;
>         RECORD_LAYER_reset_packet_length(&s->rlayer);
>         return 0;
>     }
> ```
> 
> I'm not sure what your point is with this code. It seems to me to be doing exactly what the RFC says, i.e. we have encountered an invalid record and therefore we are silently discarding it. No alert should be generated in this case (otherwise it wouldn't be silent).

Some other case will also send faltal level alert, and some seems result from invalid record. So  faltal level alert will send only in case  of internal error and only ingnore in case of invalid record as per the current design currently ,right?
`    /* r->length is now just compressed */
    if (s->expand != NULL) {
        if (rr->length > SSL3_RT_MAX_COMPRESSED_LENGTH) {
            SSLfatal(s, SSL_AD_RECORD_OVERFLOW, SSL_F_DTLS1_PROCESS_RECORD,
                     SSL_R_COMPRESSED_LENGTH_TOO_LONG);
            return 0;
        }
        if (!ssl3_do_uncompress(s, rr)) {
            SSLfatal(s, SSL_AD_DECOMPRESSION_FAILURE,
                     SSL_F_DTLS1_PROCESS_RECORD, SSL_R_BAD_DECOMPRESSION);
            return 0;
        }
    }
`

**Comment by KanPlus** [comment]
> 

Thanks, As per the RFC6347, Maybe it is better to ingnore in case of invalid record and send faltal level alert in case of internal error.

**Comment by mattcaswell** [comment]
> Thanks, As per the RFC6347, Maybe it is better to ingnore in case of invalid record and send faltal level alert in case of internal error.

Yes - that is the intent - but it seems there are a couple of cases as I mentioned above that need to be fixed.

**Comment by KanPlus** [comment]
> 

Thanks, I will check further and raise PR to fix

**Comment by nhorman** [comment]
marking as inactive, will be closed at the end of the 3.4 dev cycle barring further input
